src/twisted/mail/smtp.py

```diff
@@ -1923,21 +1928,34 @@ def clientConnectionLost(self, connector, err):
         self._processConnectionError(connector, err)
 
     def _processConnectionError(self, connector, err):
+        # If the initial connection succeeded, but there was a later error
+        # (such as TLS validation failure), we're more interested in that than
+        # "Connection was closed cleanly" or whatever.
+        if self.currentProtocol and self.currentProtocol.failureReason:
```

Comment:
1. We don't want to test `self.currentProtocol.__bool__` here, we just want to check if it `is not None`
2. I find that inline ternaries are more expressive than overloading boolean logic for tests like this, so, for clarity, I think I'd prefer this:

```suggestion
        if self.currentProtocol.failureReason is not None if self.currentProtocol is not None else False:
```

The first point is semantic, although the ternary thing is just stylistic, so feel free to skip it if you'd prefer `… is not None and … is not None`

Reply:
Honestly I find the ternary expression utterly incomprehensible, but I've corrected this to check for Noneness.